app/controllers/declaration/SealController.scala

```diff
@@ -0,0 +1,112 @@
+import play.api.mvc.{Action, AnyContent}
+import services.CustomsCacheService
+import uk.gov.hmrc.play.bootstrap.controller.FrontendController
+import views.html.declaration.seal
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class SealController @Inject()(
+  override val messagesApi: MessagesApi,
+  authenticate: AuthAction,
+  journeyType: JourneyAction,
+  errorHandler: ErrorHandler,
+  cacheService: CustomsCacheService
+)(implicit ec: ExecutionContext, appConfig: AppConfig)
+    extends FrontendController with I18nSupport {
+
+  def form(): Form[Seal] = Form(formMapping)
+  val sealsAllowed = 9999
+
+  def displayForm(): Action[AnyContent] = (authenticate andThen journeyType).async { implicit request =>
+    cacheService
+      .fetchAndGetEntry[Seq[Seal]](cacheId, formId)
+      .map { data =>
+        Ok(seal(form, data.getOrElse(Seq.empty)))
+      }
+  }
+
+  def submitForm(): Action[AnyContent] = (authenticate andThen journeyType).async { implicit request =>
+    val actionTypeOpt = request.body.asFormUrlEncoded.map(FormAction.fromUrlEncoded(_))
+
+    cacheService
+      .fetchAndGetEntry[Seq[Seal]](cacheId, formId)
+      .flatMap(cache => processRequest(cache.getOrElse(Seq.empty), actionTypeOpt))
+  }
+
+  private def processRequest(cachedSeals: Seq[Seal], action: Option[FormAction])(
+    implicit request: JourneyRequest[_]
+  ) = {
+    val boundForm = form.bindFromRequest()
+    action match {
+      case Some(Add) => addSeal(boundForm, sealsAllowed, cachedSeals)
+
+      case Some(Remove(ids)) => removeSeal(cachedSeals, ids)
+
+      case Some(SaveAndContinue) => saveSeal(boundForm, sealsAllowed, cachedSeals)
+
+      case _ => errorHandler.displayErrorPage()
+    }
+  }
+
+  private def saveSeal(boundForm: Form[Seal], elementLimit: Int, cachedSeals: Seq[Seal])(
+    implicit request: JourneyRequest[_],
+    appConfig: AppConfig
+  ) =
+    saveAndContinue(boundForm, cachedSeals, false, elementLimit).fold(
+      formWithErrors => Future.successful(BadRequest(seal(formWithErrors, cachedSeals))),
+      updatedCache =>
+        if (updatedCache != cachedSeals)
+          cacheService
+            .cache[Seq[Seal]](cacheId, formId, updatedCache)
+            .map(_ => Redirect(SummaryPageController.displayPage()))
+        else Future.successful(Redirect(SummaryPageController.displayPage()))
+    )
+
+  private def removeSeal(cachedSeals: Seq[Seal], ids: Seq[String])(implicit request: JourneyRequest[_]) =
+    redirect(remove(ids.headOption, cachedSeals))
+
+
+  private def addSeal(boundForm: Form[Seal], elementLimit: Int, seals: Seq[Seal])(
+    implicit request: JourneyRequest[_],
+    appConfig: AppConfig
+  ) =
+    add(boundForm, seals, elementLimit).fold(
+      formWithErrors => Future.successful(BadRequest(seal(formWithErrors, seals))),
+      updatedCache => redirect(updatedCache)
+    )
+
+  private def redirect(seals: Seq[Seal])(implicit request: JourneyRequest[_]) = cacheService
```

Comment:
Please change name to `cacheAndRedirect`

redirect is misleading

Reply:
updated